File: Sources/KalshiQuickSearchApp/Models/MarketModels.swift
import Foundation

enum MarketOutcomeSide: String, Codable, Sendable {
    case yes
    case no
}

enum MatchField: String, Codable, Sendable {
    case title
    case subtitle
    case outcome
    case eventTitle
    case description
    case category
    case none
}

struct TextHighlightRange: Hashable, Codable, Sendable {
    let start: Int
    let length: Int
}

struct MarketSummary: Identifiable, Hashable, Codable, Sendable {
    let ticker: String
    let eventTicker: String?
    let seriesTicker: String?
    let title: String
    let subtitle: String?
    let yesLabel: String?
    let noLabel: String?
    let eventTitle: String?
    let eventSubtitle: String?
    let description: String?
    let category: String?
    let status: String
    let lastPrice: Double?
    let yesPrice: Double?
    let noPrice: Double?
    let volume: Double?
    let volume24h: Double?
    let openInterest: Double?
    let liquidity: Double?
    let updatedAt: Date?
    let openTime: Date?
    let closeTime: Date?
    let imageURL: URL?
    let webURL: URL?

    var id: String { ticker }

    var resolvedWebURL: URL {
        if let webURL {
            return webURL
        }

        if let eventTicker {
            let encodedEvent = eventTicker.lowercased()
            let encodedMarket = ticker.lowercased()
            if let url = URL(string: "https://kalshi.com/markets/\(encodedEvent)/\(encodedMarket)") {
                return url
            }
        }

        if let url = URL(string: "https://kalshi.com/browse?query=\(ticker.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ticker)") {
            return url
        }

        return URL(string: "https://kalshi.com")!
    }

    var displayOdds: Int? {
        let price = yesPrice ?? lastPrice
        guard let price else { return nil }
        return Int((price * 100).rounded())
    }

    var volumeSignal: Double {
        max(volume24h ?? 0, volume ?? 0)
    }
}

struct SearchResult: Identifiable, Hashable, Sendable {
    let market: MarketSummary
    let score: Double
    let matchedField: MatchField
    let matchedOutcome: MarketOutcomeSide?
    let titleHighlights: [TextHighlightRange]
    let subtitleHighlights: [TextHighlightRange]
    let outcomeHighlights: [TextHighlightRange]

    var id: String { market.id }

    var emphasizedOdds: Int? {
        switch matchedOutcome {
        case .yes:
            guard let value = market.yesPrice else { return market.displayOdds }
            return Int((value * 100).rounded())
        case .no:
            guard let value = market.noPrice else { return market.displayOdds }
            return Int((value * 100).rounded())
        case .none:
            return market.displayOdds
        }
    }

    var emphasizedOutcomeLabel: String {
        switch matchedOutcome {
        case .yes:
            return market.yesLabel ?? "YES"
        case .no:
            return market.noLabel ?? "NO"
        case .none:
            return "YES"
        }
    }
}

File: packages/android/app/src/main/java/com/truthpulse/data/MarketModels.kt
package com.truthpulse.data

import kotlinx.serialization.Serializable
import java.net.URLEncoder
import kotlin.math.roundToInt

enum class MarketOutcomeSide { YES, NO }

enum class MatchField { TITLE, SUBTITLE, OUTCOME, EVENT_TITLE, DESCRIPTION, CATEGORY, NONE }

data class TextHighlightRange(val start: Int, val length: Int)

@Serializable
data class MarketSummary(
    val ticker: String,
    val eventTicker: String? = null,
    val seriesTicker: String? = null,
    val title: String,
    val subtitle: String? = null,
    val yesLabel: String? = null,
    val noLabel: String? = null,
    val eventTitle: String? = null,
    val eventSubtitle: String? = null,
    val category: String? = null,
    val status: String = "active",
    val lastPrice: Double? = null,
    val yesPrice: Double? = null,
    val noPrice: Double? = null,
    val volume: Double? = null,
    val volume24h: Double? = null,
    val openInterest: Double? = null,
    val liquidity: Double? = null,
    val updatedAtMillis: Long? = null,
    val openTimeMillis: Long? = null,
    val closeTimeMillis: Long? = null,
    val imageUrl: String? = null,
    val webUrl: String? = null,
) {
    val id: String get() = ticker

    val resolvedWebUrl: String
        get() {
            webUrl?.let { return it }

            if (seriesTicker != null && eventTicker != null) {
                val series = seriesTicker.lowercase()
                val event = eventTicker.lowercase()
                return "https://kalshi.com/markets/$series/m/$event"
            }

            val encoded = URLEncoder.encode(title, "UTF-8")
            return "https://kalshi.com/browse?query=$encoded"
        }

    val displayOdds: Int?
        get() {
            val price = yesPrice ?: lastPrice ?: return null
            return (price * 100).roundToInt()
        }

    val volumeSignal: Double
        get() = maxOf(volume24h ?: 0.0, volume ?: 0.0)
}

data class SearchResult(
    val market: MarketSummary,
    val score: Double,
    val matchedField: MatchField,
    val matchedOutcome: MarketOutcomeSide?,
    val titleHighlights: List<TextHighlightRange>,
    val subtitleHighlights: List<TextHighlightRange>,
    val outcomeHighlights: List<TextHighlightRange>,
) {
    val id: String get() = market.id

    val emphasizedOdds: Int?
        get() = when (matchedOutcome) {
            MarketOutcomeSide.YES -> {
                market.yesPrice?.let { (it * 100).roundToInt() } ?: market.displayOdds
            }
            MarketOutcomeSide.NO -> {
                market.noPrice?.let { (it * 100).roundToInt() } ?: market.displayOdds
            }
            null -> market.displayOdds
        }

    val emphasizedOutcomeLabel: String
        get() = when (matchedOutcome) {
            MarketOutcomeSide.YES -> market.yesLabel ?: "YES"
            MarketOutcomeSide.NO -> market.noLabel ?: "NO"
            null -> "YES"
        }
}
